File: py/bclab.py
'''
Created on 2017/09/12

@author: admin
'''

import paho.mqtt.client as mqtt
import json
import datetime
import traceback
import socket

TOPIC_STATUS = "bcbus/status"
TOPIC_COMMAND = "bcbus/command"
TOPIC_DATA = "bcbus/data"
TOPIC_EVENT = "bcbus/event"
TOPIC_ANALYSIS = "bcbus/analysis"

STATUS_DISCONNECTED = "STATE_DISCONNECTED"
STATUS_CONNECTED = "STATE_CONNECTED"
STATUS_ARMED = "STATE_ARMED"
STATUS_TRIGGERED = "STATE_TRIGGERED"

COMMAND_CONNECT = "COMMAND_CONNECT"
COMMAND_DISCONNECT = "COMMAND_DISCONNECT"
COMMAND_ARM = "COMMAND_ARM"
COMMAND_DISARM = "COMMAND_DISARM"
COMMAND_TRIGGER_ON = "COMMAND_TRIGGER_ON"
COMMAND_TRIGGER_OFF = "COMMAND_TRIGGER_OFF"
COMMAND_STATUS = "COMMAND_STATUS"
COMMAND_MARK = "COMMAND_MARK"

QOS = 1



class DataChannel():
    def __init__(self, name, offset, range_min, range_max, signal_type):
        self.name = name
        self.offset = offset
        self.range_min = range_min
        self.range_max = range_max
        self.signal_type = signal_type
                 
class Component:
    _status = STATUS_DISCONNECTED
    _status_message = ''
    
    state_transfers = {
        STATUS_DISCONNECTED: {COMMAND_CONNECT: STATUS_CONNECTED },
        STATUS_CONNECTED: {COMMAND_DISCONNECT: STATUS_DISCONNECTED, COMMAND_ARM: STATUS_ARMED},
        STATUS_ARMED: {COMMAND_TRIGGER_ON: STATUS_TRIGGERED, COMMAND_DISARM: STATUS_CONNECTED},
        STATUS_TRIGGERED: {COMMAND_TRIGGER_OFF: STATUS_CONNECTED}
    }

    def __init__(self, logger, broker_address, broker_port, client_id):
        logger.info('"%s" connecting to: %s:%s' % (client_id, broker_address , broker_port))
        self.log = logger
        self.client_id = client_id
       
        self.mq = mqtt.Client(client_id, clean_session=True, userdata=self, protocol=mqtt.MQTTv31)
 
        self.mq.on_message = self._on_message
        self.mq.on_log = self._on_log
        self.mq.on_connect = self._on_connect

        self.mq.connect(broker_address, int(broker_port), keepalive=60)

        self.mq.subscribe(TOPIC_COMMAND, qos=QOS)
        self.mq.subscribe(TOPIC_EVENT, qos=QOS)
   
        self.notify_status()
        self.mq.loop_start()
    
    def _on_connect(self, mq, connector, flags, rc):
        if rc:
            connector.log.error("Connector returned " + mqtt.error_string(rc))
        else:
            connector.log.info("Connected to Broker")
            connector.notify_status()
            
    def _on_message(self, mq,  connector, msg):
        try:
            msg.payload = msg.payload.decode('utf-8')
            payload = json.loads(msg.payload)
    
            if msg.topic == TOPIC_COMMAND:
                connector.on_command(payload)
            elif msg.topic == TOPIC_DATA:
                if 'data' in payload:
                    if connector.get_status() in [STATUS_ARMED, STATUS_TRIGGERED]:
                        connector.on_data_stream(msg.topic, payload)
                else:
                    connector.on_data_init(msg.topic, payload)
            elif 'analysis_id' in payload:
                connector.on_analysis(msg.topic, payload)
            elif msg.topic == TOPIC_EVENT:
                connector.on_event(msg.topic, payload)
            elif msg.topic == TOPIC_STATUS:
                connector.on_status(payload)
    
        except:
            err_msg = traceback.format_exc()
            connector.log.error(err_msg)
            connector._status_message = err_msg
            connector.notify_status()


    def _on_log(self, mq, connector, level, string):
        if level == mqtt.MQTT_LOG_ERR:
            connector.log.error(string)
        elif level == mqtt.MQTT_LOG_WARNING:
            connector.log.warning(string)
    
        
    def get_status(self):
        return self._status

    def on_command(self, msg):
        command = msg['command']        

        self.log.info("COMMAND: " + command + " " + str(msg))
        valid_transfers = self.state_transfers[self._status]
        if command in valid_transfers:
            if command == COMMAND_CONNECT:
                self.on_connect(msg)
            elif command == COMMAND_ARM:
                self.on_arm(msg)
            elif command == COMMAND_TRIGGER_ON:
                self.on_trigger_on(msg)
            elif command == COMMAND_TRIGGER_OFF:
                self.on_trigger_off(msg)
            elif command == COMMAND_DISARM:
                self.on_disarm(msg)
            elif command == COMMAND_DISCONNECT:
                self.on_disconnect(msg)
                
            new_status = valid_transfers[command]
            self._status = new_status
            self.log.info( "status changed to " + new_status)
            self.notify_status()            
        elif command == COMMAND_STATUS:
            self.on_status(msg)
            self.notify_status()
                    
    def notify_status(self):
        payload = {"status": self._status,
                   "client_id": self.client_id,
                   "time_utc": datetime.datetime.utcnow().isoformat(),
                   "clock_id": socket.gethostname(),
                   "message": self._status_message}
        (result, _) = self.mq.publish(TOPIC_STATUS, json.dumps(payload), qos=QOS)
        if result:
            self.log.error(mqtt.error_string(result))
            
            
    def subscribe_data(self, topic_id):
        (result, _) = self.mq.subscribe(topic_id, qos=QOS)
        if result:
            self.log.error(self.mq.error_string(result))
        
    def subscribe_analysis(self, analysis_id):
        (result, _) = self.mq.subscribe(analysis_id, qos=QOS)
        if result:
            self.log.error(self.mq.error_string(result))

    def post_analysis(self, analysis_id, **kwargs):
        kwargs['analysis_id'] = analysis_id
        self.log.info('post analysis')
        self.log.info(kwargs)
        (result, _) = self.mq.publish(analysis_id, json.dumps(kwargs), qos=QOS)
        if result:
            self.log.error(self.mq.error_string(result))

    def post_event(self, event_id, **kwargs):
        kwargs['event_id'] = event_id
        self.log.info('post event')
        self.log.info(kwargs)
        (result, _) = self.mq.publish(TOPIC_EVENT, json.dumps(kwargs), qos=QOS)     
        if result:
            self.log.error(self.mq.error_string(result))
  

    def __del__(self):
        self.mq.loop_stop()
        self.mq.disconnect()
        self.mq = None
        self.listeners = []
        
    def on_data_init(self, channel_id, msg):
        pass
    
    def on_data_stream(self, channel_id, msg):
        pass
    
    def on_analysis(self, analysis_id, msg):
        pass
        
    def on_event(self, event_id, msg):
        pass
    
    def on_connect(self, attr):
        pass
    
    def on_arm(self, attr):
        pass

    def on_trigger_on(self, attr):
        pass
    
    def on_trigger_off(self, attr):
        pass

    def on_disarm(self, attr):
        pass
    
    def on_disconnect(self, attr):
        pass       
    
    def on_status(self, attr):
        pass
    
class Producer(Component):
    def __init__(self, logger, broker_address, broker_port, client_id):
        super(Producer, self).__init__(logger, broker_address, broker_port, client_id)
        
        self.stream_id = ''
        self.stream_initializer = ""
        
    def on_status(self, attr):
        if  self.stream_initializer: 
            self._send_stream_initializer()
            
    def initialize_data(self, sampling_rate, data_stride, channels):
        self.stream_id = str(datetime.datetime.now())
        ch_out = []
        for ch in channels:
            ch_out.append(ch.__dict__)
        self.log.info(json.dumps(ch_out))

        payload = {"client_id": self.client_id,
                   "time_utc":datetime.datetime.utcnow().isoformat(), 
                   "clock_id": socket.gethostname(),
                   "sampling_rate": sampling_rate, 
                   "data_stride": data_stride, 
                   "channels": ch_out, 
                   "stream_id": self.stream_id}
        
        self.stream_initializer = json.dumps(payload)
        self._send_stream_initializer()

    def _send_stream_initializer(self):
        (result, _mid) = self.mq.publish(TOPIC_DATA, self.stream_initializer, qos=QOS)
        if result:
            self.log.error(self.mq.error_string(result))


    def stream_data(self, data_string, first_scan, scan_count):
        payload = {"client_id": self.client_id, 
                   "first_scan": first_scan, 
                   "scan_count": scan_count, 
                   "data": data_string, 
                   "stream_id": self.stream_id}
        (result, _mid) = self.mq.publish(TOPIC_DATA, json.dumps(payload), qos=QOS)
        if result:
            self.log.error(self.mq.error_string(result))

class Controller(Component):   
    def __init__(self, logger, broker_address, broker_port, client_id):
        super(Controller, self).__init__(logger, broker_address, broker_port, client_id)
        self.mq.subscribe(TOPIC_STATUS, qos=QOS)
    
    def on_status(self, attr):
        pass
         
    def command(self, command, **kwargs):
        kwargs['command'] = command
        self.mq.publish(TOPIC_COMMAND, json.dumps(kwargs), qos=QOS)
    
    def connect(self):
        self.command(COMMAND_CONNECT)
        
    def arm(self, dst_path=''):
        self.command(COMMAND_ARM, dst_path=dst_path)
        
    def trigger_on(self):
        self.command(COMMAND_TRIGGER_ON)
        
    def trigger_off(self):
        self.command(COMMAND_TRIGGER_OFF)
    
    def disarm(self):    
        self.command(COMMAND_DISARM)
    
    def disconnect(self):
        self.command(COMMAND_DISCONNECT)

File: web/js/bclab.js

var _busConnector; 

var TOPIC_COMMAND = "bcbus/command";
var TOPIC_STATUS = "bcbus/status";
var TOPIC_EVENT = "bcbus/event";


var COMMAND_CONNECT = "COMMAND_CONNECT";
var COMMAND_ARM = "COMMAND_ARM";
var COMMAND_TRIGGER_ON = "COMMAND_TRIGGER_ON";
var COMMAND_TRIGGER_OFF = "COMMAND_TRIGGER_OFF";
var COMMAND_DISARM = "COMMAND_DISARM";
var COMMAND_DISCONNECT = "COMMAND_DISCONNECT";
var COMMAND_STATUS = "COMMAND_STATUS";

function BcbusConnector(brokerAddr, brokerPort, plots, analysis, statusListeners, eventListeners) {
	this.STATE_DISCONNECTED = "STATE_DISCONNECTED";
	this.STATE_CONNECTED = "STATE_CONNECTED";
	this.STATE_ARMED = "STATE_ARMED";
	this.STATE_TRIGGERED = "STATE_TRIGGERED";

    _busConnector = this;

	var clientName = "bc" + new Date().getTime().toString(10);
    this.client = new Messaging.Client(brokerAddr, brokerPort, clientName);
    console.log("BcbusConnector Client id = " + clientName + "Connecting to " + this.client._getHost() + ":" + this.client._getPort());
    this.client.startTrace();


    this.plots = plots;
    this.analysis = analysis;
    this.statusListeners = statusListeners;
    this.eventListeners = eventListeners;
    this.client.onMessageArrived = this.handleMessage;
    this.client.connect({onSuccess:this.handleConnect, onFailure:this.handleFailure});
    console.log("BcbusConnector created");
}

BcbusConnector.prototype.handleConnect = function() {
	var subs = [TOPIC_STATUS,TOPIC_EVENT]
	for (p in _busConnector.plots) {
		subs.push(_busConnector.plots[p].client)
	}
	for (a in _busConnector.analysis) {
		subs.push(_busConnector.analysis[a].analysis_id)
	}
	for (s in subs) {
		_busConnector.client.subscribe(subs[s])
	}
	_busConnector.status()
	console.log("BcbusConnector connected");
};

BcbusConnector.prototype.handleFailure = function(obj) {
	console.log(obj.errorMessage);
};

BcbusConnector.prototype.handleMessage = function(message) {
	try {
		var payload = JSON.parse(message.payloadString);
		switch (message.destinationName) {
			case TOPIC_STATUS:
				_busConnector.processStatus(message);
				break;
	
			case TOPIC_EVENT:
				_busConnector.processEvents(message)
				break;
		}
		if (payload.hasOwnProperty('stream_id')) {
			_busConnector.processData(payload);
		} else if (payload.hasOwnProperty('analysis_id')) {
			_busConnector.processAnalysis(payload);
		}


	
	} catch (exp) {
		console.log("Exception at message handler:" + exp);
	}
};

BcbusConnector.prototype.processData = function(payload) {
	

	var client = payload.client_id;
	if ('channels' in payload ) {
		var initializer =  payload;
		var signals = [];
		for (var ci = 0; ci < initializer.channels.length; ++ci) {
			var channel = initializer.channels[ci];
			var signal = new Signal(channel.name, initializer.data_stride, channel.offset, channel.range_min, channel.range_max );
			signals.push(signal);
		}

        _busConnector.plots.forEach(function(plot) {
            if ((!plot.client) || (client == plot.client)) {
                plot.calibrate(initializer.stream_id, signals, signals.length, 1, initializer.sampling_rate );
            }
        });
	}
	if ('data' in payload) {
	    var byteArray = base64js.toByteArray(payload.data);

		var data = new Float32Array(byteArray.buffer)
        _busConnector.plots.forEach(function(plot) {
            if ((!plot.client) || (client == plot.client)) {
                plot.draw(payload.stream_id, data, payload.first_scan, payload.scan_count);
            }
        });
	}
};

BcbusConnector.prototype.processAnalysis = function(payload) {

	var client = payload.client_name;

    _busConnector.analysis.forEach(function(ana) {
		try {
			var patt=new RegExp(ana.analysis_name)
				if ( patt.test( payload.name)) {
					payload.analysis.name=payload.name;
					ana.draw(payload.analysis);
				}
		} catch(exp) {
				console.log("Exception at analysis:" + exp);
		}
    });

};

BcbusConnector.prototype.processStatus= function(message) {
	if (typeof _busConnector.statusListeners === 'undefined') {
		return;
	}
	
	var payload = JSON.parse(message.payloadString);
	var client = payload.client_id;

    _busConnector.statusListeners.forEach(function(el) {
    	try {
            el.update(payload);
        } catch(exp) {
        	console.log("Exception at status:" + exp);
		}
    });	
};

BcbusConnector.prototype.processEvents = function(message) {
	if (typeof _busConnector.eventListeners === 'undefined') {
		return;
	}
	var payload = JSON.parse(message.payloadString);
	var client = payload.client_name;

    _busConnector.eventListeners.forEach(function(el) {
 try {
            el.update(payload);
            } catch(exp) {
		console.log("Exception at events:" + exp);

}
    });
};

BcbusConnector.prototype.send = function(cmd) {
	var msg = new Messaging.Message(cmd);
	msg.destinationName = TOPIC_COMMAND;
	_busConnector.client.send(msg);

};

BcbusConnector.prototype.send_event = function(event) {
	var json = JSON.stringify(event);
	var msg = new Messaging.Message(json);
	msg.destinationName = TOPIC_EVENT;
	_busConnector.client.send(msg);
}

BcbusConnector.prototype.connect = function() {
	var cmd = JSON.stringify({command:COMMAND_CONNECT})
	_busConnector.send(cmd);
};

BcbusConnector.prototype.arm = function(dest) {
	var msg = {command: COMMAND_ARM, destination: dest};
	var json = JSON.stringify(msg);
	_busConnector.send(json);
};

BcbusConnector.prototype.triggerOn = function() {
	var cmd = JSON.stringify({command:COMMAND_TRIGGER_ON})
	_busConnector.send(cmd);
};

BcbusConnector.prototype.triggerOff = function() {
	var cmd = JSON.stringify({command:COMMAND_TRIGGER_OFF})
	_busConnector.send(cmd);
};

BcbusConnector.prototype.disarm = function() {
	var cmd = JSON.stringify({command:COMMAND_DISARM})
	_busConnector.send(cmd);
};

BcbusConnector.prototype.disconnect = function() {
	var cmd = JSON.stringify({command:COMMAND_DISCONNECT})
	_busConnector.send(cmd);
};

BcbusConnector.prototype.status = function() {
	var cmd = JSON.stringify({command:COMMAND_STATUS})
	_busConnector.send(cmd);
};


